File: Assets/Scripts/Scene/AI/AITree/AITreeImpleted.cs
using UnityEngine;
using System.Collections.Generic;
using AI.Runtime;

namespace AI
{
    public class AITreeImpleted 
    {
        public static XEntity Transform2Entity(Transform t)
        {
            uint id = uint.Parse(t.name);
            return XEntityMgr.singleton.GetEntity(id);
        }

        public static AIRuntimeStatus NavToTargetUpdate(XEntity entity,GameObject mAIArgTarget,GameObject mAIArgNavTarget,Vector3 mAIArgNavPos)
        {
            if (!XEntity.Valide(entity)) return AIRuntimeStatus.Failure;
            if (mAIArgTarget == null)
            {
                if (mAIArgNavTarget == null)
                {
                    if (mAIArgNavPos == Vector3.zero)
                        return AIRuntimeStatus.Failure;
                    else
                    {
                        if (ActionNav(entity, mAIArgNavPos))
                            return AIRuntimeStatus.Success;
                        else
                            return AIRuntimeStatus.Failure;
                    }
                }
                else
                {
                    if (NavToTarget(entity, mAIArgNavTarget))
                        return AIRuntimeStatus.Success;
                    else
                        return AIRuntimeStatus.Failure;
                }
            }
            else
            {
                if (NavToTarget(entity, mAIArgTarget))
                    return AIRuntimeStatus.Success;
                else
                    return AIRuntimeStatus.Failure;
            }
        }
        
        public static AIRuntimeStatus StopNavMoveUpdate(XEntity entity)
        {
            if (XEntity.Valide(entity) && StopNavMove(entity))
                return AIRuntimeStatus.Success;
            return AIRuntimeStatus.Failure;
        }

        public static AIRuntimeStatus RotateToTargetUpdate(XEntity entity, GameObject mAIArgTarget, float ang)
        {
            if (XEntity.Valide(entity) && RotateToTarget(entity.EntityTransfer, mAIArgTarget, ang))
                return AIRuntimeStatus.Success;
            else
                return AIRuntimeStatus.Failure;
        }

        public static AIRuntimeStatus RotateSelfUpdate(XEntity entity, float max,float min)
        {
            if (XEntity.Valide(entity))
            {
                float ang = Random.Range(min, max);
                Vector3 dir = XCommon.singleton.HorizontalRotateVetor3(entity.Forward, ang);
                entity.EntityTransfer.forward = dir;
                return AIRuntimeStatus.Success;
            }
            return AIRuntimeStatus.Failure;
        }

        public static AIRuntimeStatus MoveForwardUpdate(XEntity entity)
        {
            if (XEntity.Valide(entity))
            {
                entity.MoveForward(entity.EntityTransfer.forward);
                return AIRuntimeStatus.Success;
            }
            else
                return AIRuntimeStatus.Failure;
        }

        public static AIRuntimeStatus DetectEnemyInSightUpdate(XEntity entity)
        {
            if (XEntity.Valide(entity) && DetectEnemyInSight(entity))
                return AIRuntimeStatus.Success;
            else
                return AIRuntimeStatus.Failure;
        }


        public static AIRuntimeStatus FindTargetByDistanceUpdate(XEntity entity,float mAIArgDistance,float mAIArgAngle)
        {
            if (XEntity.Valide(entity) && entity.GetComponent<XAIComponent>().FindTargetByDistance(mAIArgDistance, mAIArgAngle))
                return AIRuntimeStatus.Success;
            else
                return AIRuntimeStatus.Failure;
        }


        public static AIRuntimeStatus ResetTargetUpdate(XEntity entity)
        {
            if (XEntity.Valide(entity) && entity.GetComponent<XAIComponent>().ResetTarget())
                return AIRuntimeStatus.Success;
            return AIRuntimeStatus.Failure;
        }

        public static AIRuntimeStatus DoSelectNearestUpdate(XEntity entity)
        {
            if (XEntity.Valide(entity) && entity.GetComponent<XAIComponent>().DoSelectNearest())
                return AIRuntimeStatus.Success;
            else
                return AIRuntimeStatus.Failure;
        }


        public static AIRuntimeStatus DoSelectFarthestUpdate(XEntity entity)
        {
            if (XEntity.Valide(entity) && entity.GetComponent<XAIComponent>().DoSelectFarthest())
                return AIRuntimeStatus.Success;
            else
                return AIRuntimeStatus.Failure;
        }

        public static AIRuntimeStatus DoSelectRandomTargetUpdate(XEntity entity)
        {
            if (entity != null && entity.GetComponent<XAIComponent>().DoSelectRandom())
                return AIRuntimeStatus.Success;
            else
                return AIRuntimeStatus.Failure;
        }


        public static AIRuntimeStatus CalDistanceUpdate(XEntity entity, Transform mAIArgObject, float mAIArgDistance, Vector3 mAIArgDestPoint)
        {
            if (!XEntity.Valide(entity)) return AIRuntimeStatus.Failure;
            if (mAIArgObject != null)
            {
                mAIArgDistance = (entity.Position - mAIArgObject.position).magnitude;
            }
            else
            {
                mAIArgDistance = (entity.Position - mAIArgDestPoint).magnitude;
            }
            return AIRuntimeStatus.Success;
        }

        public static AIRuntimeStatus SelectMoveTargetByIdUpdate(XEntity entity, Transform mAIArgMoveTarget, int mAIArgObjectId)
        {
            if (!XEntity.Valide(entity)) return AIRuntimeStatus.Failure;
            Transform moveTarget = SelectMoveTargetById(entity, mAIArgObjectId);
            if (moveTarget == null)
                return AIRuntimeStatus.Failure;
            else
            {
                mAIArgMoveTarget = moveTarget;
                return AIRuntimeStatus.Success;
            }
        }

        public static AIRuntimeStatus ValueHPUpdate(XEntity entity, int mAIArgMaxHP,int mAIArgMinHP)
        {
            if (!XEntity.Valide(entity)) return AIRuntimeStatus.Failure;
            double hp = entity.Attributes.GetAttr(XAttributeDefine.XAttr_CurrentHP_Basic);
            if (hp >= mAIArgMinHP && hp <= mAIArgMaxHP)
                return AIRuntimeStatus.Success;
            else
                return AIRuntimeStatus.Failure;
        }

        public static AIRuntimeStatus ValueMPUpdate(XEntity entity, int mAIArgMaxMP, int mAIArgMinMP)
        {
            if (!XEntity.Valide(entity)) return AIRuntimeStatus.Failure;
            double hp = entity.Attributes.GetAttr(XAttributeDefine.XAttr_CurrentMP_Basic);
            if (hp >= mAIArgMinMP && hp <= mAIArgMaxMP)
                return AIRuntimeStatus.Success;
            else
                return AIRuntimeStatus.Failure;
        }

        public static AIRuntimeStatus ValueTargetUpdate(XEntity entity, Transform mAIArgTarget)
        {
            if (XEntity.Valide(mAIArgTarget))
                return AIRuntimeStatus.Success;
            else
            {
                if (XEntity.Valide(entity))
                {
                    entity.GetComponent<XAIComponent>().SetTarget(null);
                }
                return AIRuntimeStatus.Failure;
            }
        }

        public static AIRuntimeStatus ValueDistanceUpdate(XEntity entity, GameObject mAIArgTarget, float mAIArgMaxDistance)
        {
            if (mAIArgTarget == null)
                return AIRuntimeStatus.Failure;

            float dis = (entity.Position - mAIArgTarget.transform.position).sqrMagnitude;
            if (dis <= mAIArgMaxDistance * mAIArgMaxDistance)
                return AIRuntimeStatus.Success;
            else
                return AIRuntimeStatus.Failure;
        }


        public static AIRuntimeStatus StatusRandomUpdate(XEntity e, int prob)
        {
            if (prob < 0 || prob > 100)
            {
                return AIRuntimeStatus.Failure;
            }
            else
            {
                int rand = Random.Range(0, 100);
                if (rand >= prob)
                    return AIRuntimeStatus.Success;
                else return AIRuntimeStatus.Failure;
            }
        }

        public static AIRuntimeStatus IsOppoCastingSkillUpdate(XEntity entity)
        {
            if (!XEntity.Valide(entity)) return AIRuntimeStatus.Failure;
            if (entity.GetComponent<XAIComponent>().IsOppoCastingSkill)
                return AIRuntimeStatus.Success;
            else
                return AIRuntimeStatus.Failure;
        }


        public static AIRuntimeStatus IsHurtOppoUpdate(XEntity entity)
        {
            if (!XEntity.Valide(entity)) return AIRuntimeStatus.Failure;
            if (entity.GetComponent<XAIComponent>().IsHurtOppo)
                return AIRuntimeStatus.Success;
            else
                return AIRuntimeStatus.Failure;
        }

        public static AIRuntimeStatus IsFixedInCdUpdate(XEntity entity)
        {
            if (!XEntity.Valide(entity)) return AIRuntimeStatus.Failure;
            if (entity.GetComponent<XAIComponent>().IsFixedInCd)
                return AIRuntimeStatus.Success;
            else
                return AIRuntimeStatus.Failure;
        }

        public static AIRuntimeStatus IsCastingSkillUpdate(XEntity entity)
        {
            if (!XEntity.Valide(entity)) return AIRuntimeStatus.Failure;
            if (entity.GetComponent<XAIComponent>().IsCastingSkill)
                return AIRuntimeStatus.Success;
            else
                return AIRuntimeStatus.Failure;
        }

        public static AIRuntimeStatus StatusIdleUpdate(XEntity entity)
        {
            if (!XEntity.Valide(entity)) return AIRuntimeStatus.Failure;

            if (entity.CurState == XStateDefine.XState_Idle)
                return AIRuntimeStatus.Success;
            else
                return AIRuntimeStatus.Failure;
        }

        public static AIRuntimeStatus StatusMoveUpdate(XEntity entity)
        {
            if (!XEntity.Valide(entity)) return AIRuntimeStatus.Failure;

            if (entity.CurState == XStateDefine.XState_Move)
                return AIRuntimeStatus.Success;
            else
                return AIRuntimeStatus.Failure;
        }

        public static AIRuntimeStatus StatusBehitUpdate(XEntity entity)
        {
            if (!XEntity.Valide(entity)) return AIRuntimeStatus.Failure;

            if (entity.CurState == XStateDefine.XState_BeHit)
                return AIRuntimeStatus.Success;
            else
                return AIRuntimeStatus.Failure;
        }


        public static AIRuntimeStatus StatusDeathUpdate(XEntity entity)
        {
            if (!XEntity.Valide(entity)) return AIRuntimeStatus.Failure;

            if (entity.CurState == XStateDefine.XState_Death)
                return AIRuntimeStatus.Success;
            else
                return AIRuntimeStatus.Failure;
        }

        public static AIRuntimeStatus StatusFreezeUpdate(XEntity entity)
        {
            if (!XEntity.Valide(entity)) return AIRuntimeStatus.Failure;

            if (entity.CurState == XStateDefine.XState_Freeze)
                return AIRuntimeStatus.Success;
            else
                return AIRuntimeStatus.Failure;
        }

        public static AIRuntimeStatus StatusSkillUpdate(XEntity entity)
        {
            if (!XEntity.Valide(entity)) return AIRuntimeStatus.Failure;

            if (entity.CurState == XStateDefine.XState_Skill)
                return AIRuntimeStatus.Success;
            else
                return AIRuntimeStatus.Failure;
        }

        public static AIRuntimeStatus LogUpdate(XEntity entity, string str, bool b)
        {
            if (b) XDebug.LogError(str);
            else XDebug.Log(str);
            return AIRuntimeStatus.Success;
        }

        public static AIRuntimeStatus WaitUpdate(XEntity entity, float wait, bool rand, float min, float max)
        {
            if (entity.SetTimer(wait)) return AIRuntimeStatus.Success;
            return AIRuntimeStatus.Failure;
        }


        public static AIRuntimeStatus XAIActionSkillUpdate(XEntity entity, string scr, GameObject target)
        {
            var skill = entity.GetComponent<XSkillComponent>();
            if (skill != null && target != null)
            {
                if (!skill.IsCasting)
                    skill.CastSkill(scr);
                else
                    return AIRuntimeStatus.Running;
            }
            return AIRuntimeStatus.Failure;
        }


        public static AIRuntimeStatus RandomCompareUpdate(XEntity entity, int prob)
        {
            int rand = Random.Range(0, 100);
            if (prob > rand)
            {
                return AIRuntimeStatus.Success;
            }
            else
            {
                return AIRuntimeStatus.Failure;
            }
        }


        public static Transform SelectMoveTargetById(XEntity entity, int objectid)
        {
            List<XEntity> ens = XEntityMgr.singleton.GetAllEnemy(entity);
            for (int i = 0, max = ens.Count; i < max; i++)
            {
                if (XEntity.Valide(ens[i]) && ens[i].Attributes.TypeID == objectid)
                {
                    return ens[i].EntityObject.transform;
                }
            }
            return null;
        }

        public static bool ActionNav(XEntity entity, Vector3 dest)
        {
            XNavComponent nav = entity.GetComponent<XNavComponent>();
            if (nav != null)
            {
                nav.Navigate(dest);
                entity.MoveForward(dest - entity.Position);
                return true;
            }
            return false;
        }


        public static bool NavToTarget(XEntity entity, GameObject target)
        {
            if (entity == null) return false;
            if (target != null)
            {
                XNavComponent nav = entity.GetComponent<XNavComponent>();
                if (nav != null)
                {
                    nav.Navigate(target.transform.position);
                    entity.MoveForward(target.transform.position - entity.Position);
                    return true;
                }
            }
            return false;
        }

        public static bool StopNavMove(XEntity entity)
        {
            if (entity == null) return false;
            XNavComponent nav = entity.GetComponent<XNavComponent>();
            if (nav == null) return false;
            nav.NavEnd();
            return true;
        }

        public static bool RotateToTarget(Transform src, GameObject target,float ang)
        {
            if (src != null && target != null)
            {
                Vector3 dir = target.transform.position - src.position;
                dir = XCommon.singleton.HorizontalRotateVetor3(dir, ang);
                src.forward = dir;
                return true;
            }
            return false;
        }

        public static bool DetectEnemyInSight(XEntity e)
        {
            List<XEntity> opponent = XEntityMgr.singleton.GetAllEnemy(e);
            for (int i = 0; i < opponent.Count; i++)
            {
                if (!XEntity.Valide(opponent[i])) continue;
                Vector3 dir = opponent[i].Position - e.Position;
                float distance = dir.sqrMagnitude;
                //一旦在视野范围，就激活仇恨列表
                if (distance < e.Attributes.EnterFightRange * e.Attributes.EnterFightRange)
                {
                    XAIComponent ai = opponent[i].GetComponent<XAIComponent>();
                    ai.SetTarget(e);
                    return true;
                }
            }
            return false;
        }

    }
}

File: tools_proj/XCPP/GameCore/ai/AITreeImpleted.cpp
#include "AITreeImpleted.h"
#include "../XEntityMgr.h"
#include "../Transform.h"
#include "../XEntity.h"
#include "../XAIComponent.h"
#include "../XAttributes.h"
#include "../XNavigationComponent.h"

#define CHECKENTY(entity) if (!XEntity::Valide(entity)) return Failure;


XEntity* AITreeImpleted::Transform2Entity(Transform* t)
{
	uint id = 0;
	std::stringstream ss(t->name);
	ss >> id;
	return XEntityMgr::Instance()->GetEntity(id);
}

AIStatus AITreeImpleted::NavToTargetUpdate(XEntity* entity, GameObject* mAIArgTarget, GameObject* mAIArgNavTarget, Vector3 mAIArgNavPos)
{
	CHECKENTY(entity);
	if (mAIArgTarget)
	{
		if (NavToTarget(entity, mAIArgTarget))
			return Success;
		else
			return Failure;
	}
	else
	{
		if (mAIArgNavTarget)
		{
			if (NavToTarget(entity, mAIArgNavTarget))
				return Success;
			else
				return Failure;
		}
		else
		{
			if (mAIArgNavPos == Vector3::zero)
				return Failure;
			else
			{
				if (ActionNav(entity, mAIArgNavPos))
					return Success;
				else
					return Failure;
			}
		}
	}
}

AIStatus AITreeImpleted::StopNavMoveUpdate(XEntity* entity)
{
	CHECKENTY(entity);
	if (StopNavMove(entity)) return Success;
	return Failure;
}

AIStatus AITreeImpleted::RotateToTargetUpdate(XEntity* entity, GameObject* mAIArgTarget, float ang)
{
	CHECKENTY(entity);
	if (RotateToTarget(entity->getTransfer(), mAIArgTarget, ang)) return Success;
	return Failure;
}

AIStatus AITreeImpleted::RotateSelfUpdate(XEntity* entity, float max, float min)
{
	CHECKENTY(entity);
	float ang = Random((int)(min * 100), (int)(max * 100)) / 100.0f;
	//Vector3 dir = XCommon.singleton.HorizontalRotateVetor3(entity.Forward, ang);
	//entity.EntityTransfer.forward = dir;
	return Success;
}

AIStatus AITreeImpleted::MoveForwardUpdate(XEntity* entity)
{
	CHECKENTY(entity);
	entity->MoveForward(entity->getTransfer()->forward);
	return Success;
}

AIStatus AITreeImpleted::DetectEnemyInSightUpdate(XEntity* entity)
{
	CHECKENTY(entity);
	if (DetectEnemyInSight(entity))
		return Success;
	return Failure;
}

AIStatus AITreeImpleted::FindTargetByDistanceUpdate(XEntity* entity, float mAIArgDistance, float mAIArgAngle)
{
	CHECKENTY(entity);
	if (entity->GetComponent<XAIComponent>()->FindTargetByDistance(mAIArgDistance, mAIArgAngle))
	{
		return Success;
	}
	return Failure;
}

AIStatus AITreeImpleted::ResetTargetUpdate(XEntity* entity)
{
	CHECKENTY(entity);
	if (entity->GetComponent<XAIComponent>()->ResetTarget())
		return Success;
	return Failure;
}

AIStatus AITreeImpleted::DoSelectNearestUpdate(XEntity* entity)
{
	CHECKENTY(entity);
	if (entity->GetComponent<XAIComponent>()->DoSelectNearest())
		return Success;
	else
		return Failure;
}

AIStatus AITreeImpleted::DoSelectFarthestUpdate(XEntity* entity)
{
	CHECKENTY(entity);
	if (entity->GetComponent<XAIComponent>()->DoSelectFarthest())
		return Success;
	else
		return Failure;
}

AIStatus AITreeImpleted::DoSelectRandomTargetUpdate(XEntity* entity)
{
	CHECKENTY(entity);
	if (entity->GetComponent<XAIComponent>()->DoSelectRandom())
		return Success;
	else
		return Failure;
}

AIStatus AITreeImpleted::CalDistanceUpdate(XEntity* entity, Transform* mAIArgObject, float mAIArgDistance, Vector3 mAIArgDestPoint)
{
	CHECKENTY(entity);
	if (mAIArgObject)
	{
		Vector3 v = entity->getPostion() - mAIArgObject->position;
		mAIArgDistance = v.magnitude();
	}
	else
	{
		mAIArgDistance = (entity->getPostion() - mAIArgDestPoint).magnitude();
	}
	return Success;
}

AIStatus AITreeImpleted::SelectMoveTargetByIdUpdate(XEntity* entity, Transform* mAIArgMoveTarget, int mAIArgObjectId)
{
	CHECKENTY(entity);
	Transform* moveTarget = SelectMoveTargetById(entity, mAIArgObjectId);
	if (moveTarget)
	{
		mAIArgMoveTarget = moveTarget;
		return Success;
	}
	return Failure;
}

AIStatus AITreeImpleted::ValueHPUpdate(XEntity* entity, int mAIArgMaxHP, int mAIArgMinHP)
{
	CHECKENTY(entity);
	double hp = entity->getAttributes()->GetAttr(XAttr_CurrentHP_Basic);
	if (hp >= mAIArgMinHP && hp <= mAIArgMaxHP)
		return Success;
	else
		return Failure;
}

AIStatus AITreeImpleted::ValueMPUpdate(XEntity* entity, int mAIArgMaxMP, int mAIArgMinMP)
{
	CHECKENTY(entity);
	double hp = entity->getAttributes()->GetAttr(XAttr_CurrentMP_Basic);
	if (hp >= mAIArgMinMP && hp <= mAIArgMaxMP)
		return Success;
	else
		return Failure;
}

AIStatus AITreeImpleted::ValueTargetUpdate(XEntity* entity, Transform* mAIArgTarget)
{
	CHECKENTY(entity);
	return Success;
}

AIStatus AITreeImpleted::ValueDistanceUpdate(XEntity* entity, GameObject* mAIArgTarget, float mAIArgMaxDistance)
{
	CHECKENTY(entity);
	float dis = Vector3::sqrtMagnitude(entity->getPostion() - mAIArgTarget->transform->position);
	if (mAIArgTarget &&dis <= mAIArgMaxDistance * mAIArgMaxDistance)
		return Success;
	else
		return Failure;
}

AIStatus AITreeImpleted::StatusRandomUpdate(XEntity* entity, int prob)
{
	CHECKENTY(entity);
	if (prob < 0 || prob > 100)
	{
		return Failure;
	}
	else
	{
		int randx = Random(0, 100);
		if (randx >= prob)
			return Success;
		else return Failure;
	}
}

AIStatus AITreeImpleted::IsOppoCastingSkillUpdate(XEntity* entity)
{
	CHECKENTY(entity);
	if (entity->GetComponent<XAIComponent>()->getIsOppoCastingSkill())
		return Success;
	else
		return Failure;
}

AIStatus AITreeImpleted::IsHurtOppoUpdate(XEntity* entity)
{
	CHECKENTY(entity);
	if (entity->GetComponent<XAIComponent>()->getIsHurtOppo())
		return Success;
	else
		return Failure;
}

AIStatus AITreeImpleted::IsFixedInCdUpdate(XEntity* entity)
{
	CHECKENTY(entity);
	if (entity->GetComponent<XAIComponent>()->getIsFixedInCd())
		return Success;
	else
		return Failure;
}

AIStatus AITreeImpleted::IsCastingSkillUpdate(XEntity* entity)
{
	CHECKENTY(entity);
	if (entity->GetComponent<XAIComponent>()->getIsCastingSkill())
		return Success;
	else
		return Failure;
}

AIStatus AITreeImpleted::StatusIdleUpdate(XEntity* entity)
{
	CHECKENTY(entity);
	if (entity->getState() == XState_Idle)
		return Success;
	else
		return Failure;
}

AIStatus AITreeImpleted::StatusMoveUpdate(XEntity* entity)
{
	CHECKENTY(entity);
	if (entity->getState() == XState_Move)
		return Success;
	else
		return Failure;
}

AIStatus AITreeImpleted::StatusBehitUpdate(XEntity* entity)
{
	CHECKENTY(entity);
	if (entity->getState() == XState_BeHit)
		return Success;
	else
		return Failure;
}

AIStatus AITreeImpleted::StatusDeathUpdate(XEntity* entity)
{
	CHECKENTY(entity);
	if (entity->getState() == XState_Death)
		return Success;
	else
		return Failure;
}

AIStatus AITreeImpleted::StatusFreezeUpdate(XEntity* entity)
{
	CHECKENTY(entity);
	if (entity->getState() == XState_Freeze)
		return Success;
	else
		return Failure;
}

AIStatus AITreeImpleted::StatusSkillUpdate(XEntity* entity)
{
	CHECKENTY(entity);
	if (entity->getState() == XState_Skill)
		return Success;
	else
		return Failure;
}

AIStatus AITreeImpleted::LogUpdate(XEntity* entity, std::string str, bool b)
{
	if (b) ERR(str);
	else LOG(str);
	return Success;
}

AIStatus AITreeImpleted::WaitUpdate(XEntity* entity, float wait, bool rand, float min, float max)
{
	CHECKENTY(entity);
	return Success;

}

AIStatus AITreeImpleted::XAIActionSkillUpdate(XEntity* entity, std::string scr, GameObject* target)
{
	CHECKENTY(entity);
	return Success;
}

AIStatus AITreeImpleted::RandomCompareUpdate(XEntity* entity, int prob)
{
	CHECKENTY(entity);
	int randx = Random(0,100);
	if (prob > randx)
	{
		return Success;
	}
	else
	{
		return Failure;
	}
}

Transform* AITreeImpleted::SelectMoveTargetById(XEntity* entity, int objectid)
{
	std::vector<XEntity*> ens = XEntityMgr::Instance()->GetAllEnemy(entity);
	for (size_t i = 0, max = ens.size(); i < max; i++)
	{
		if (XEntity::Valide(ens[i]) && ens[i]->getAttributes()->getTypeID() == objectid)
		{
			return ens[i]->getTransfer();
		}
	}
	return 0;
}

bool AITreeImpleted::ActionNav(XEntity* entity, Vector3 dest)
{
	XNavigationComponent* nav = entity->GetComponent<XNavigationComponent>();
	if (nav)
	{
		nav->Navigate(dest);
		entity->MoveForward(dest - entity->getPostion());
		return true;
	}
	return false;
}

bool AITreeImpleted::NavToTarget(XEntity* entity, GameObject* target)
{
	if (target)
	{
		XNavigationComponent* nav = entity->GetComponent<XNavigationComponent>();
		if (nav)
		{
			nav->Navigate(target->transform->position);
			entity->MoveForward(target->transform->position - entity->getPostion());
			return true;
		}
	}
	return false;
}

bool AITreeImpleted::StopNavMove(XEntity* entity)
{
	if (entity == NULL) return false;
	XNavigationComponent* nav = entity->GetComponent<XNavigationComponent>();
	if (nav) nav->NavEnd();
	return true;
}

bool AITreeImpleted::RotateToTarget(Transform* src, GameObject* target, float ang)
{
	if (src  && target)
	{
		Vector3 dir = target->transform->position - src->position;
		//dir = XCommon.singleton.HorizontalRotateVetor3(dir, ang);
		src->forward = dir;
		return true;
	}
	return false;
}

bool AITreeImpleted::DetectEnemyInSight(XEntity* e)
{
	std::vector<XEntity*> opponent = XEntityMgr::Instance()->GetAllEnemy(e);
	for (int i = 0; i < opponent.size(); i++)
	{
		if (!XEntity::Valide(opponent[i])) continue;
		Vector3 dir = opponent[i]->getPostion() - e->getPostion();
		float distance = Vector3::sqrtMagnitude(dir);
		//一旦在视野范围，就激活仇恨列表
		if (distance < e->getAttributes()->EnterFightRange * e->getAttributes()->EnterFightRange)
		{
			XAIComponent* ai = opponent[i]->GetComponent<XAIComponent>();
			ai->SetTarget(e);
			return true;
		}
	}
	return false;
}